### ISSUE TITLE
Deduping across files

### ISSUE THREAD

**Issue by mstade** [postcss/postcss-import#303]
In #297 the core issue seems to be that @hirokiosame wants to `@import` the same dependency into multiple *different* files. I have a similar issue – in their case, it's the want to reuse font declarations across multiple files but *not* embed them. In my case, the need is for variables.

If my files were *purely* variables, this wouldn't be much of an issue, since by using postcss-next they'll get inlined. However, if I *weren't* using postcss-next, or if those files include other selectors than `:root { }`, I'll get the same content repeated multiple times. Not only is this wasteful, it also runs the risk of applying selectors in the wrong order.

I'm *not* using webpack, so filtering certain files won't work unfortunately.

At first I was thinking I could just make a PR and introduce a "global" state map for storing the hashes of imported files, to essentially mimic the per-file deduping behaviour that postcss-import already has, but then I started thinking whether this would be enough, considering order matters. Consider the following tree:

```
A
├─ C
│  └╴B
├─ D 
│  └╴B
└─ B
```

In this, `A` is kind of a main file which imports `C`, `D`, and `B` in that order. In turn, `C` and `D` imports `B` again. Without global deduping, this means `B` is imported three times, and twice in incorrect positions. A flattened file will be `A + C + B + D + B + B`, when it should probably be `A + B + C + D`. The reason it should be `A + B + C + D` rather than `A + C + D + B` is of course that the `@import` rule must come first in files, and so should be loaded *before* the file depending on it. As well, it probably makes sense to hoist things as high as possible. I think if you have a dependency tree up front, the order is the same as a depth first search through the tree.

The above scenario assumes an entry point, but there may be no such thing. The entry point may just be the order in which things are fed to the postcss. In my case, I'm using the JS API to process files, but I keep the same processor for all files. In essence, I have something like this:

```
const files = getFilesToBuild()
const cc = postcss([cssimport, ... otherPlugins])
const builds = files.map(file => cc.process(file))
const results = await Promise.all(builds)
```

In this case, `results` would be an array of all the build results from postcss. Ignoring obvious issues like keeping all files in memory and so on, would a global hashing as described above be sufficient to implement this kind of deduping? Is state kept across the life time of a postcss processor instance, or is it kept only while processing a single file? Would there be a risk of race conditions, where for instance `D` is processed before `C` and therefor the order may be messed up after all?

I'm keen to raise a PR to implement this kind of deduping, but before diving in head first I was hoping maybe I could pick your collective brains a bit. :o)

**Comment by RyanZim** [comment]
@mstade postcss-import already has deduping for the scenario with an entry point you described. The output would be `A + B + C + D`.

However, deduping state is only kept for the lifetime of processing the entry point (i.e. a single `process()` call). I'm not sure how/why you'd want deduping across multiple entry points?

**Comment by mstade** [comment]
Then I must be doing something wrong, because I'm most definitely seeing duplication. I thought this must have been the case at first, but probably I've made a mistake somewhere in my debugging. I will investigate more thoroughly and try to figure it out.

With regards to your question – it's certainly not unthinkable that you may have multiple entry points (in fact it's pretty common in our environments) however if things should work as you say, then really it should just be a matter of doing a `process()` call *per* entry point.

My apologies for (probably) wasting your time. I'll investigate further, and close this ticket for now. Many thanks for your insight @RyanZim!

**Comment by RyanZim** [comment]
If you get stuck debugging, hop in https://gitter.im/postcss/postcss; I'm around there and there are plenty of other helpful people there too.

**Comment by mstade** [comment]
Hey @RyanZim – thanks a bunch for the offer and the assistance! I managed to track down the issue and indeed, it was my fault. Turns out there was a rogue copy operation that would overwrite the outputted file when I tried the entry point approach, so it was broken, but the output was suspiciously similar to what the actual output should have been, just with a bunch of dupes in there as well.

When I noticed this, thanks to your nudge, postcss-import worked just as advertised. Many thanks for your help and my sincere apologies for wasting your time. 😊
